cou/steps/plan.py

```diff
@@ -129,13 +129,26 @@ async def _generate_data_plane_plan(
     hypervisor_apps, non_hypervisors_apps = _separate_hypervisors_apps(
         analysis_result.apps_data_plane
     )
+    plans = []
+    if args.upgrade_group != HYPERVISORS:
+        plans.extend(
+            _generate_data_plane_subordinate_remaining_plan(
+                target, non_hypervisors_apps, args.force
+            )
+        )
 
-    plans = [
-        await _generate_data_plane_hypervisors_plan(target, analysis_result, args, hypervisor_apps)
-    ]
+    plans.extend(
+        [
+            await _generate_data_plane_hypervisors_plan(
+                target, analysis_result, args, hypervisor_apps
+            )
+        ]
+    )
 
     if args.upgrade_group != HYPERVISORS:
-        plans.extend(_generate_data_plane_remaining_plan(target, non_hypervisors_apps, args.force))
+        plans.extend(
+            _generate_data_plane_principle_remaining_plan(target, non_hypervisors_apps, args.force)
+        )
```

Comment:
I think it should be like

```python3
plans = []
(
  non_hypervisor_subordinate, 
  non_hypervisor_principle, 
  hypervisors_subordinate, 
  hypervisors_principle
)  = _separate_hypervisors_apps(analysis_result.apps_data_plane)

if args.upgrade_group != HYPERVISORS:
    plans.extend(
        _generate_data_plane_non_hypervisor_subordinate_plan(
            target, non_hypervisor_subordinate, args.force
        )
    )
    plans.extend(
        _generate_data_plane_non_hypervisor_principle_plan(target, non_hypervisor_principle, args.force)
    )
plans.extend(
    [
        await _generate_data_plane_hypervisors_subordinate_plan(
            target, analysis_result, args, hypervisors_subordinate
        ),
        await _generate_data_plane_hypervisors_principle_plan(
            target, analysis_result, args, hypervisors_principle
        )
    ]
)
```

Reply:
Thanks Ray.

We have to be very careful about this one. Maybe we should discuss every applications upgrade strategy here. 

For example, ceph-osd is a data plan principle. If you move `ceph-osd` to `non_hypervisor_principle`, it will break the upgrade order.`ceph-osd` should upgrade after `nova-compute`.

Most of the applications doesn't support single unit upgrade as I understand. Which means on hypervisor upgrading strategy, we will break the recommend upgrade order. For example you upgrade nova then ceph-osd on machine 1, but ceph-osd will be refresh at the same time on machine 2 which nova still no upgraded yet.

Maybe we should define the whole plan first? Like below is a example:

```
non_hypervisor_subordinate
non_hypervisor_principle_before_nova
hypervisor_unit_by_unit_upgrade
  unit_upgrade_supported_principle, before nova
  nova_upgrade
  unit_upgrade_supported_principle, after nova
non_hypervisor_principle_after_nova
```

cc: @gabrielcocenza

Reply:
As mentioned by James, non-hypervisors should upgrade `after nova-compute` because of ceph-osd. Note that even if ceph-osd is deployed in the same nova-compute machines we consider as a non-hypervisor because it does not have the ability to upgrade unit-by-unit.

We are already upgrading applications by sequence, meaning that if cinder is deployed with nova-compute, we will upgrade cinder first and then nova-compute. Note that you need to disable nova-compute scheduler before upgrading cinder to avoid new instances in a hypervisor while the upgrade is happening.

In summary I do believe that the current approach is solid, the only thing is that we need to upgrade hypervisors-subordinates first.

- hypervisors_subordinate
- hypervisors(including other principles E.g: cinder) respecting the current srtucture
- -non_hypervisors_subordinates
- non_hypervisors

As far as I understand is just a question of separating subordinates from hypervisors and non-hypervisors and respect the sequence. Note:
- subordinates of other principals like cinder will be considered as hypervisors_subordinate
- nova-compute subordinates are considered as hypervisors_subordinate
- ceph-osd subordinates are considered as non_hypervisors_subordinates
- if ceph-osd and nova-compute share a same subordinate we can consider as hypervisors_subordinate